src/sage/rings/polynomial/multi_polynomial.pyx

```diff
@@ -2683,7 +2683,7 @@ cdef class MPolynomial(CommutativePolynomial):
 
         if gens:
             # substituting all variables (in a polynomial ring with variables) with 0
-            d = {str(gen): 0 for gen in gens}
+            d = {str(gen): self.base_ring().zero() for gen in gens}
```

Comment:
First, you should use `self.base().zero()`; second, you also need to make the same change on line 2699 below.